### PR TITLE
Try always multiplying director tensors.

### PR DESCRIPTION
**Pull request by jperryhouts**
For speed comparison with #327

### CODE REVIEW

**Comment by jperryhouts** [comment]
Um. This is somehow faster that the comparison method now. I have no idea what to think of that... And my test case is failing, so maybe I just broke it somehow? :/ I'll play with it more tomorrow.

**Comment by jperryhouts** [comment]
Oh, the test was failing because of a bug in the anisotropic shear heating, fixed by #655 

It looks like the times (for the ctest models) seem to be comparable between the two methods (random variations, probably just because of other system load, make some difference).

However, the "always-multiply" method slows down by a factor of ~2 relative to the "comparison" method in Debug mode... For some tests, but not all.

**Comment by jperryhouts** [comment]
For example: In Debug mode, with "compare" method:
```
  1/264 Test   #1: adiabatic_conditions ...........................................   Passed   23.59 sec
  2/264 Test   #2: adiabatic_heating ..............................................   Passed   23.29 sec
  3/264 Test   #3: adiabatic_heating_simplified ...................................   Passed   22.91 sec
  4/264 Test   #4: adiabatic_initial_conditions ...................................   Passed   23.57 sec
  5/264 Test   #5: adiabatic_initial_conditions_chunk .............................***Failed    7.75 sec
  6/264 Test   #6: adiabatic_initial_conditions_chunk_3d ..........................***Failed  154.42 sec
  7/264 Test   #7: adiabatic_initial_conditions_constant ..........................   Passed   23.99 sec
  8/264 Test   #8: adiabatic_initial_conditions_subadiabaticity ...................   Passed   23.33 sec
  9/264 Test   #9: adiabatic_initial_conditions_without_boundary ..................   Passed    6.57 sec
 10/264 Test  #10: advect_mesh_vertically .........................................***Failed   24.83 sec
 11/264 Test  #11: always_refine ..................................................   Passed    6.34 sec
 12/264 Test  #12: anisotropic_viscosity ..........................................   Passed   16.24 sec
 13/264 Test  #13: artificial_postprocess .........................................   Passed   12.06 sec
...
```
And with "multiply" method:
```
  1/264 Test   #1: adiabatic_conditions ...........................................   Passed   19.01 sec
  2/264 Test   #2: adiabatic_heating ..............................................   Passed   11.88 sec
  3/264 Test   #3: adiabatic_heating_simplified ...................................   Passed   11.80 sec
  4/264 Test   #4: adiabatic_initial_conditions ...................................   Passed   18.94 sec
  5/264 Test   #5: adiabatic_initial_conditions_chunk .............................***Failed    6.60 sec
  6/264 Test   #6: adiabatic_initial_conditions_chunk_3d ..........................***Failed   78.58 sec
  7/264 Test   #7: adiabatic_initial_conditions_constant ..........................   Passed   19.06 sec
  8/264 Test   #8: adiabatic_initial_conditions_subadiabaticity ...................   Passed   18.89 sec
  9/264 Test   #9: adiabatic_initial_conditions_without_boundary ..................   Passed    5.29 sec
 10/264 Test  #10: advect_mesh_vertically .........................................***Failed   15.15 sec
 11/264 Test  #11: always_refine ..................................................   Passed    5.36 sec
 12/264 Test  #12: anisotropic_viscosity ..........................................   Passed    8.49 sec
 13/264 Test  #13: artificial_postprocess .........................................   Passed    4.74 sec
...
```

**Comment by jperryhouts** [comment]
I ran all the tests for both the debug and release mode in each of the cases (the comparison version, the multiplication version, and the original version: with no director tensors at all). Here are the timing results:
```
debug-multiply.tests:      Total Test time (real) = 1247.56 sec
debug-nodirection.tests:   Total Test time (real) = 791.78 sec
debug-compare.tests:       Total Test time (real) = 788.17 sec
release-multiply.tests:    Total Test time (real) = 497.04 sec
release-nodirection.tests: Total Test time (real) = 484.43 sec
release-compare.tests:     Total Test time (real) = 487.22 sec
```

**Comment by bangerth** [comment]
I finally found some time and ran a modification of shell-simple-3d on my system, and here is what I have:

Current master, debug mode: 
```
| Assemble Stokes system          |         8 |       121s |        28% |
| Assemble temperature system     |         8 |       156s |        36% |
| Build Stokes preconditioner     |         3 |      47.2s |        11% |
| Build temperature preconditioner|         8 |     0.551s |      0.13% |
| Solve Stokes system             |         8 |      38.2s |       8.9% |
```

Current master, release mode:
```
| Assemble Stokes system          |         8 |      5.96s |         7% |
| Assemble temperature system     |         8 |      5.38s |       6.3% |
| Build Stokes preconditioner     |         3 |      6.38s |       7.5% |
| Build temperature preconditioner|         8 |     0.564s |      0.66% |
| Solve Stokes system             |         8 |      62.6s |        74% |
```


With your patch, debug mode:
```
| Assemble Stokes system          |         8 |       811s |        63% |
| Assemble temperature system     |         8 |       160s |        12% |
| Build Stokes preconditioner     |         3 |       199s |        15% |
| Build temperature preconditioner|         8 |     0.565s |     0.044% |
| Solve Stokes system             |         8 |      50.3s |       3.9% |
```

With your patch, release mode:
```
| Assemble Stokes system          |         8 |      22.2s |        23% |
| Assemble temperature system     |         8 |      5.16s |       5.4% |
| Build Stokes preconditioner     |         3 |      8.31s |       8.6% |
| Build temperature preconditioner|         8 |     0.542s |      0.56% |
| Solve Stokes system             |         8 |      55.9s |        58% |
```

It's not quite clear to me what is going on with the different solver times (maybe something else was running on the system), but in either case, the new assembly is incurring a sizable performance hit, even in release mode. This is with the current version of the pull request.

This was in 3d, where I expect the difference to be largest. I suspect that the tests mostly run in 2d, so maybe the difference isn't so visible there.

Would you mind repeating your tests with a 3d testcase and with all 6 versions you had? I used `cookbook/shell_simple_3d.prm` with the following changes:
```
diff --git a/cookbooks/shell_simple_3d.prm b/cookbooks/shell_simple_3d.prm
index 1a26209..fbad434 100644
--- a/cookbooks/shell_simple_3d.prm
+++ b/cookbooks/shell_simple_3d.prm
@@ -4,7 +4,7 @@
 
 set Dimension                              = 3
 set Use years in output instead of seconds = true
-set End time                               = 1.5e9
+set End time                               = 1.5e6
 set Output directory                       = output
 
 
@@ -62,15 +62,15 @@ end
 
 
 subsection Mesh refinement
-  set Initial global refinement          = 2
-  set Initial adaptive refinement        = 3
+  set Initial global refinement          = 1
+  set Initial adaptive refinement        = 2
   set Strategy                           = temperature
   set Time steps between mesh refinement = 15
 end
 
 
 subsection Postprocess
-  set List of postprocessors = visualization, velocity statistics, temperature statistics, heat flux statistics, depth average
+  set List of postprocessors =
 
   subsection Visualization
     set Output format                 = vtu
```
With 8 processors, this should run in a few minutes (around 7 minutes in debug mode, about 1.5 minutes in release mode).

**Comment by jperryhouts** [comment]
Here they are in Release mode:
```
==> build/master-release/log.txt <==
| Total wallclock time elapsed since start    |      96.8s |            |
| Assemble Stokes system          |         8 |      6.53s |       6.7% |
| Assemble temperature system     |         8 |       6.3s |       6.5% |
| Build Stokes preconditioner     |         3 |       5.9s |       6.1% |
| Build temperature preconditioner|         8 |     0.598s |      0.62% |
| Solve Stokes system             |         8 |      72.2s |        75% |
==> build/multiply-release/log.txt <==
| Total wallclock time elapsed since start    |       108s |            |
| Assemble Stokes system          |         8 |      17.6s |        16% |
| Assemble temperature system     |         8 |      6.36s |       5.9% |
| Build Stokes preconditioner     |         3 |      8.07s |       7.5% |
| Build temperature preconditioner|         8 |     0.577s |      0.53% |
| Solve Stokes system             |         8 |      70.4s |        65% |
==> build/no-direction-release/log.txt <==
| Total wallclock time elapsed since start    |      94.4s |            |
| Assemble Stokes system          |         8 |       6.1s |       6.5% |
| Assemble temperature system     |         8 |      6.35s |       6.7% |
| Build Stokes preconditioner     |         3 |      5.35s |       5.7% |
| Build temperature preconditioner|         8 |     0.682s |      0.72% |
| Solve Stokes system             |         8 |      70.5s |        75% |
```
And in Debug mode:
```
==> build/master-debug/log.txt <==
| Total wallclock time elapsed since start    |       227s |            |
| Assemble Stokes system          |         8 |      51.8s |        23% |
| Assemble temperature system     |         8 |      62.4s |        28% |
| Build Stokes preconditioner     |         3 |        17s |       7.5% |
| Build temperature preconditioner|         8 |     0.613s |      0.27% |
| Solve Stokes system             |         8 |      70.8s |        31% |
==> build/multiply-debug/log.txt <==
| Total wallclock time elapsed since start    |       596s |            |
| Assemble Stokes system          |         8 |       358s |        60% |
| Assemble temperature system     |         8 |      59.8s |        10% |
| Build Stokes preconditioner     |         3 |      87.1s |        15% |
| Build temperature preconditioner|         8 |     0.566s |     0.095% |
| Solve Stokes system             |         8 |      67.8s |        11% |
==> build/no-direction-debug/log.txt <==
| Total wallclock time elapsed since start    |       222s |            |
| Assemble Stokes system          |         8 |      50.3s |        23% |
| Assemble temperature system     |         8 |      58.5s |        26% |
| Build Stokes preconditioner     |         3 |      22.3s |        10% |
| Build temperature preconditioner|         8 |     0.593s |      0.27% |
| Solve Stokes system             |         8 |      68.6s |        31% |
```

I assumed there must be lots of Asserts in the `operator *` functions for SymmetricTensors, but it looks like that isn't the case. The only substantial difference I can find is that multiplying tensors (in the `perform_double_contraction` functions) requires allocating some temporary memory and copying things around, and maybe Debug mode inherently performs more checks on these than Release mode does?

**Comment by bangerth** [comment]
Good question. Maybe we first need to address the performance problem in deal.II. The multiplication of rank-2 and rank-4 tensors ultimately ends up here:
```
 template <int dim, typename Number>
  inline
  typename SymmetricTensorAccessors::double_contraction_result<4,2,dim,Number>::type
  perform_double_contraction (const typename SymmetricTensorAccessors::StorageType<4,dim,Number>::base_tensor_type &data,
                              const typename SymmetricTensorAccessors::StorageType<2,dim,Number>::base_tensor_type &sdata)
  {
    const unsigned int data_dim =
      SymmetricTensorAccessors::StorageType<2,dim,Number>::n_independent_components;
    Number tmp [data_dim];
    for (unsigned int i=0; i<data_dim; ++i)
      tmp[i] = perform_double_contraction<dim,Number>(data[i], sdata);
    return SymmetricTensor<2,dim,Number>(tmp);
  }
```
This all looks inoccuous. All data is allocated on the stack and the compiler should be able to see through everything. We call the rank-2/rank-2 function in the loop:
```
  template <int dim, typename Number>
  inline
  typename SymmetricTensorAccessors::double_contraction_result<2,2,dim,Number>::type
  perform_double_contraction (const typename SymmetricTensorAccessors::StorageType<2,dim,Number>::base_tensor_type &data,
                              const typename SymmetricTensorAccessors::StorageType<2,dim,Number>::base_tensor_type &sdata)
  {
    switch (dim)
      {
      case 1:
        return data[0] * sdata[0];
      case 2:
        return (data[0] * sdata[0] +
                data[1] * sdata[1] +
                2*data[2] * sdata[2]);
      default:
        // Start with the non-diagonal part to avoid some multiplications by
        // 2.
        Number sum = data[dim] * sdata[dim];
        for (unsigned int d=dim+1; d<(dim*(dim+1)/2); ++d)
          sum += data[d] * sdata[d];
        sum *= 2.;
        for (unsigned int d=0; d<dim; ++d)
          sum += data[d] * sdata[d];
        return sum;
      }
  }
```
Here, `base_data_type` is simply a `Tensor<1,(dim*dim + dim)/2>` and the repeated call to `operator[]` on these objects certainly does hit assertions -- though not in release mode. I have to say I'm mystified by overhead.

Have you tried what happens in a version where you do let material models compute these tensors, but then simply ignore them (i.e., not change assembly at all)? In other words, determine whether the overhead is really in the assembly, or in computing the additional tensors? (But then, I can't imagine what difference that would make, given that the simple material model used in the testcase doesn't even know anything about these tensors, other than the fact that we need to initialize them somehow!?)

**Comment by jperryhouts** [comment]
Commenting out all the `Assert`s in tensor.h brings the worst case down to
```
| Total wallclock time elapsed since start    |       326s |            |
| Assemble Stokes system          |         8 |       156s |        48% |
| Assemble temperature system     |         8 |      41.3s |        13% |
| Build Stokes preconditioner     |         3 |      43.1s |        13% |
| Build temperature preconditioner|         8 |     0.565s |      0.17% |
| Solve Stokes system             |         8 |      68.2s |        21% |
```

**Comment by bangerth** [comment]
Which of the four cases does this correspond to?

**Comment by jperryhouts** [comment]
Oh, it's this patch, in debug mode.

**Comment by jperryhouts** [comment]
...which means over 50% of the assembly time is spent in Assert statements. Assembly is still ~3x slower than the master branch, but down from 6x.

All this is kind of moot though, I don't think any of this warrants removing those asserts just for aesthetics, but this patch is unacceptably slow otherwise. I propose just sticking with the ugly code as it is in the master branch, and if someone wants to use the director tensors they ought to do it in Release mode.

**Comment by bangerth** [comment]
I'm trying to address some of the slow contractions in https://github.com/dealii/dealii/pull/1888

**Comment by bangerth** [comment]
I ran my testcase again on this patch, and here is what I find when I try both deal.II's current master plus https://github.com/dealii/dealii/pull/2042 (="always_inline patch"), with and without the patch in this pull request:
```
DEBUG + always_inline patch
| Assemble Stokes system          |         8 |      31.4s |        26% |
| Assemble temperature system     |         8 |      41.8s |        34% |
| Build Stokes preconditioner     |         3 |      12.1s |       9.7% |
| Build temperature preconditioner|         8 |     0.194s |      0.16% |
| Solve Stokes system             |         8 |      12.4s |       9.7% |

DEBUG + always_inline patch + always multiply with tensor
| Assemble Stokes system          |         8 |       176s |        59% |
| Assemble temperature system     |         8 |      41.5s |        14% |
| Build Stokes preconditioner     |         3 |      44.6s |        15% |
| Build temperature preconditioner|         8 |     0.197s |     0.065% |
| Solve Stokes system             |         8 |      13.5s |       4.5% |



RELEASE + always_inline patch
| Assemble Stokes system          |         8 |      1.66s |       8.9% |
| Assemble temperature system     |         8 |      1.63s |       8.7% |
| Build Stokes preconditioner     |         3 |         1s |       5.3% |
| Build temperature preconditioner|         8 |      0.17s |       0.9% |
| Solve Stokes system             |         8 |      12.9s |        69% |


RELEASE + always_inline patch + always multiply with tensor
| Assemble Stokes system          |         8 |      5.74s |        23% |
| Assemble temperature system     |         8 |      1.76s |         7% |
| Build Stokes preconditioner     |         3 |      1.63s |       6.4% |
| Build temperature preconditioner|         8 |      0.17s |      0.67% |
| Solve Stokes system             |         8 |      13.4s |        57% |
```

So the patch slows down assembly by about a factor of 5.5 in debug and 3.5 in release mode. I think this is unacceptable :-( 

Now, the always_inline patch (https://github.com/dealii/dealii/pull/2042) does in reality not affect symmetric tensors at all, it only marks up functions for the `Tensor` class. So I ran the same test also with the always inline patch (https://github.com/dealii/dealii/pull/2042) and in addition the unchecked index patch (https://github.com/dealii/dealii/pull/1888). This has no effect on release mode, and this by itself is already not good; in debug mode, the slowdown is even more terrifying:
```
| Assemble Stokes system          |         8 |       563s |        72% |
| Assemble temperature system     |         8 |      49.9s |       6.4% |
| Build Stokes preconditioner     |         3 |       130s |        17% |
| Build temperature preconditioner|         8 |     0.192s |     0.025% |
| Solve Stokes system             |         8 |        13s |       1.7% |
```
In other words, https://github.com/dealii/dealii/pull/1888 doesn't just not help, it's making things worse by another factor of 3 in debug mode.

In sum, this patch doesn't seem to be the way to go with this, and consequently I'll close this for now. We'll have a chance to revisit the whole issue after modularizing the assembly process (see #695).
